k8s: Disable several CiliumEndpoint status sections by default

tgraf opened the pull request:
The PATCH'ing of the status field of the CiliumEndpoint is often the bottleneck
in scalability. However, most of the frequently changing status sections
provided are not required for operation and only provide additional visibility.

Disable several CiliumEndpoint status sections by default to improve the
scalability of the CRD-backed identity allocation mode. Introduce a flag
--endpoint-status to re-enable individual status sections.

Fixes: #10413

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10490)
<!-- Reviewable:end -->

errordeveloper commented on daemon/daemon_main.go:
```diff
@@ -288,6 +288,9 @@ func init() {
 	flags.Bool(option.EnableHealthChecking, defaults.EnableHealthChecking, "Enable connectivity health checking")
 	option.BindEnv(option.EnableHealthChecking)
 
+	flags.StringSlice(option.EndpointStatus, []string{}, "Enable additional CiliumEndpoint status features")
```
Should this include example values or at least a pointer to what those value can be?

errordeveloper commented on pkg/endpoint/endpoint_status_test.go:
```diff
@@ -48,6 +49,15 @@ type endpointGeneratorSpec struct {
 	fakeControllerManager    bool
 }
 
+type endpointStatusConfiguration map[string]bool
```
Would `map[string]struct{}` be sufficient?

tgraf replied:
Technically, yes.  I think that the `bool` type self-documents the test better as it makes it obvious that a particular feature is enabled. Given these are unit tests, there is no need to micro-optimize memory use.

tgraf committed:
k8s: Disable several CiliumEndpoint status sections by default

The PATCH'ing of the status field of the CiliumEndpoint is often the bottleneck
in scalability. However, most of the frequently changing status sections
provided are not required for operation and only provide additional visibility.

Disable several CiliumEndpoint status sections by default to improve the
scalability of the CRD-backed identity allocation mode. Introduce a flag
--endpoint-status to re-enable individual status sections.

The k8sT tests are fixed to use `cilium policy get` instead of relying
on the policy in the CiliumEndpoint resource. The CEP check in the Kafka
tests is removed entirely.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on test/k8sT/Services.go:
```diff
@@ -837,32 +836,20 @@ var _ = Describe("K8sServicesTest", func() {
-				data, err := res.Filter(`{.status.policy.egress}`)
-				ExpectWithOffset(1, err).To(BeNil(), "unable to get endpoint %s metadata", podName)
-				return data.String()
+				output, err := kubectl.LoadedPolicyInFirstAgent()
```
I think that there's a subtle difference here where previously the endpoint plumbing was tested, but now it's only the policy level plumbing that's tested. Arguably we could achieve the exact same as before using `cilium endpoint list -o json`.

That said, I'm not sure whether this particular detail is necessary to validate in these tests so I'll leave it up to your discretion whether this is worth further modifying.

tgraf replied:
Yes, there is a difference. The rationale here is that we a lot of tests already covering CIDR + label policies to endpoint plumbing combined with connectivity tests. What seemed vital is to test the correct translation of service-based policies to the relevant CIDR policies.

joestringer reviewed (approved):
LGTM, one minor note below.

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -73,6 +73,7 @@ cilium-agent [flags]
       --encrypt-node                                  Enables encrypting traffic from non-Cilium pods and host networking
       --endpoint-interface-name-prefix string         Prefix of interface name shared by all endpoints (default "lxc+")
       --endpoint-queue-size int                       size of EventQueue per-endpoint (default 25)
+      --endpoint-status strings                       Enable additional CiliumEndpoint status features (policy,health,controllers,log,state)
```
Why not an "all"?

tgraf replied:
Because most users will unlikely understand the scalability consequences of enabling all options. Given the consequences, it seems more sensible to require users to opt-in by feature.

aanm commented on pkg/option/config.go:
```diff
@@ -2411,3 +2453,23 @@ func getDefaultMonitorQueueSize(numCPU int) int {
+		EndpointStatusControllers,
+		EndpointStatusLog,
+		EndpointStatusState,
+	}
```
Does not hurt to return this by order with `sort.Strings()` or ordering them manually. before returning this slice.
